src/Workspaces/CSharp/Portable/Utilities/CSharpUseImplicitTypeHelper.cs

```diff
@@ -220,10 +210,32 @@ protected override bool ShouldAnalyzeForEachStatement(ForEachStatementSyntax for
                 declarationTypeNode,
                 SyntaxFactory.IdentifierName("var").WithTriviaFrom(declarationTypeNode).WithAdditionalAnnotations(annotation));
 
-            var newTree = tree.WithRootAndOptions(newRoot, tree.Options);
-            var newSemanticModel = semanticModel.Compilation.ReplaceSyntaxTree(tree, newTree).GetSemanticModel(newTree);
+            SemanticModel newSemanticModel;
+            if (semanticModel.IsSpeculativeSemanticModel)
```

Comment:
I think this strongly needs an explanatoin as to what's going on.  Tnx!

Reply:
Which part is unclear?

Reply:
well, basically, in the IDE code it is pretty darn rare that anything actually checks if it's in an speculative model.  So a good starting point would be why we need this check, and why the code in it is what we want here.  

I don't expect it needs to be that complex.  But it would be worth it to help make what's going on here clearer for future maintainers :)

Reply:
> So a good starting point would be why we need this check ...

You can't call any speculative APIs on a speculative semantic model. Anything that uses a speculative API needs to have obtained its own semantic model from the compilation. This applies to all uses of semantic models, so it doesn't seem like it needs to be documented everywhere speculative APIs are used. Or put another way, it's more important to document the *lack* of this check than to document the use of `IsSpeculativeSemanticModel`.

> ... why the code in it is what we want here

I'm not convinced it is. There is no API to transform a speculative semantic model to another speculative semantic model via `root.ReplaceNode(...)` or similar.